src/dockerfile.rs

```diff
@@ -0,0 +1,373 @@
+use lazy_static::lazy_static;
+use regex::{Regex, RegexSetBuilder};
+use std::{collections::HashSet, error::Error, fs, path};
+
+lazy_static! {
+    static ref FROM_DIRECTIVE: regex::Regex =
+        Regex::new(r"(?i)^FROM\s(?P<image>[^\s]+)?(?:\sAS\s)?(?P<stage>[\w+]+)?").unwrap();
+    static ref VERSION_REGEX: &'static str = r"v?([0-9]+(?:(?:\.[a-z0-9]+)|(?:-(?:kb)?[0-9]+))*)";
+    static ref TAG_REGEX: regex::RegexSet = RegexSetBuilder::new(&[
+        format!(r"^{}(?P<suffix>-[a-z0-9.\-]+)?$", *VERSION_REGEX),
+        format!(r"^(?P<prefix>[a-z0-9.\-]+-)?{}$", *VERSION_REGEX),
+        format!(
+            r"^(?P<prefix>[a-z\-]+-)?{}(?P<suffix>-[a-z\-]+)?$",
+            *VERSION_REGEX
+        ),
+    ])
+    .case_insensitive(true)
+    .build()
+    .unwrap();
+}
+
+pub struct Dockerfile {
+    pub path: String,
+    pub content: String,
+}
+
+impl Dockerfile {
+    pub fn new(path: path::PathBuf) -> Result<Self, Box<dyn Error>> {
+        let dockerfile = Self {
+            path: String::from(path.to_str().unwrap()),
```

Comment:
What's the best way to remove this `unwrap()`? Since `path.to_str()` returns an `Option` but my function returns a `Result`.

Reply:
`path.to_str()?`

the `?` is a shorthand to `unwrap()`, but you trade it for some readability and intent

Reply:
@dchambersgc Would that work?
I believe that you need to match the types i.e. `Option -> Option` and `Result -> Result`.
I may be wrong here though.

Reply:
So the `?` shorthand would get rid of the `unwrap()` call as its syntactical sugar for each other.

As we are finishing this function with`Ok(dockerfile)` then we are wrapping the success  of the `Option` and returning it as the success of the `Result`, so the flow we get is `Option` -> `Some` -> `Ok(Some)` -> `Result`

The thing that this doesn't do is handle the error case, which you would need to `match` on and handle a bit more explicitly than hope with `unwrap()`:
https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap

Reply:
Could we store `path` as a `PathBuf` on the struct? Not sure what the knock-on effects of that would be, so this might just be moving the problem around, but would be worth trying...

Reply:
My preferred way here due to being an Option is to do `path.to_str().ok_or("nooooo")?;` You basically move the Option into a Result and then use the `?` for `unwrap()`;

Reply:
you could match on the option, and convert - so in the null case you turn that into a `Err("invalid path")` or whatever else could cause `path.to_str()` to fail. Spelunking into the Rust source that would have to be a bizarre condition so probably worth representing as an error.

Reply:
The bigger question is, if this errors for some reason, do you want to panic or not? If so, just the trailing `?` or `.unwrap()` is fine.

Reply:
it seems like @jackatbancast is right. i can't use `?` to unwrap an `Option` in a function that returns a `Result`: `the trait std::error::Error is not implemented for std::option::NoneError`. i'm thinking to either store it as a `PathBuf` as @matt-thomson suggested, leave it as `unwrap` since we're probably fine to panic in the face of `None`, or pattern match and return my own custom error as @tragiclifestories pointed out..